### PR TITLE
make some services more reactive to config

### PR DESCRIPTION
**Pull request by sqs**
Partial application of https://github.com/sourcegraph/cody/pull/5221 to some services: FeatureFlagProvider, ClientStateBroadcaster, UpstreamHealthProvider, OpenTelemetryService, SentryService, etc.

The idea is to make services listen to config instead of checking it once and not updating themselves when config changes, which is bad because users need to reload the editor sometimes when they change config.

## Test plan

CI

### CODE REVIEW

File: lib/shared/src/experimentation/FeatureFlagProvider.ts [hunk 1]
```diff
@@ -194,7 +216,9 @@ export class FeatureFlagProvider {
 
     private async refreshFeatureFlags(): Promise<void> {
         return wrapInActiveSpan('FeatureFlagProvider.refreshFeatureFlags', async () => {
-            const endpoint = graphqlClient.endpoint
+            const {
+                auth: { serverEndpoint: endpoint },
+            } = await currentResolvedConfig()
```
Comment (valerybugakov) [inline lib/shared/src/experimentation/FeatureFlagProvider.ts:221]: Thoughts on having a similar to `authStatus` sync-store for the configuration? 

```ts
const { serverEndpoint: endpoint } = currentConfig().auth

File: vscode/src/main.ts [hunk 2]
```diff
@@ -316,7 +316,6 @@ async function initializeSingletons(
                     graphqlClient.setConfig(config)
                     void featureFlagProvider.refresh()
```
Comment (valerybugakov) [inline vscode/src/main.ts:317]: Can we drop this since we're now subscribed to the config? 

```suggestion
```

Comment (valerybugakov) [reply vscode/src/main.ts:317]: To ensure the correct order of operations, should we subscribe `graphqlClient` to config before `featureFlagProvider`? Otherwise, `featureFlagProvider`'s internal subscription would be triggered before the update of `graphqlClient`, and the `graphqlClient` would use the outdated server URL that is out of sync with the latest config values. 

Is this a realistic scenario? I didn't test it locally. Just thinking about edge cases based on the diff here.

File: vscode/src/services/UpstreamHealthProvider.ts [hunk 3]
```diff
@@ -23,49 +26,38 @@ const INITIAL_PING_DELAY_MS = 10 * 1000 // 10 seconds
...
-        ClientConfigurationWithAccessToken,
-        'serverEndpoint' | 'customHeaders' | 'accessToken'
-    > | null = null
     private nextTimeoutId: NodeJS.Timeout | null = null
 
+    private configSubscription: Unsubscribable
+
+    constructor() {
+        // Refresh when auth (endpoint or token) changes.
+        this.configSubscription = resolvedConfig.pipe(distinctUntilChanged()).subscribe(() => {
```
Comment (valerybugakov) [inline vscode/src/services/UpstreamHealthProvider.ts:39]: Would this be triggered on any config property change?

**Review by valerybugakov — APPROVED** [review]
Love to see it all come together!

File: lib/shared/src/experimentation/FeatureFlagProvider.ts (hunk 1 above)
Comment (sqs) [reply lib/shared/src/experimentation/FeatureFlagProvider.ts:224]: Yep, will do.

Comment (sqs) [reply lib/shared/src/experimentation/FeatureFlagProvider.ts:224]: I will probably wait on this and leave currentResolvedConfig() async as I try to eliminate more calls to it and replace them with Observable. Then I'll audit the remaining call sites and see if we can/should make it sync.

File: vscode/src/main.ts (hunk 2 above)
Comment (sqs) [reply vscode/src/main.ts:317]: Even if so, SourcegraphGraphQLAPIClient should get the latest config values when it is called. I will confirm.

File: vscode/src/services/UpstreamHealthProvider.ts (hunk 3 above)
Comment (sqs) [reply vscode/src/services/UpstreamHealthProvider.ts:39]: Yes. That is inefficient, and we can restrict it with `pluck` or `map`, but that IMO is only safe enough to do when you are directly using that value and not just using it to trigger a future refresh as it is here. So this should be refactored to work that way. Otherwise you could have a problem where you think it only needs to refresh when certain config keys change, but actually the code is using other config keys and so you get staleness.

**Comment by sqs** [comment]
`test-integration (Windows)` is failing, but it's also failing on main: https://github.com/sourcegraph/cody/commits/main. Will merge this and chat with people here in Montreal who might be overhauling these this week.
